docs/repository_agents.md

```diff
@@ -0,0 +1,168 @@
+## Implementing a Repository Agent
+
+A repository agent must be implemented as a shared library and the
+name of the shared library must be
+*libtritonrepoagent_\<repo-agent-name\>.so*. The shared library should
+hide all symbols except those needed by the repository agent API. See
+the [checksum example's
+CMakeList.txt](https://github.com/triton-inference-server/checksum_repository_agent/blob/main/CMakeLists.txt)
+for an example of how to use an ldscript to expose only the necessary
+symbols.
+
+The shared library will be dynamically loaded by Triton when it is
+needed. For a repository agent called *A*, the shared library must be
+installed as /opt/tritonserver/repoagents/A/libtritonrepoagent_A.so.
+
+Your repository agent must implement the repository agent API as
+documented in
+[tritonrepoagent.h](https://github.com/triton-inference-server/core/tree/main/include/triton/core/tritonrepoagent.h).
+
+Triton follows these steps when loading a model:
+
+* Load the model's configuration file (config.pbtxt) and extract the
+  *ModelRepositoryAgents* settings. Even if a repository agent
+  modifies the config.pbtxt file, the repository agent settings from
+  the initial config.pbtxt file are used for the entire loading
+  process.
+
+* For each repository agent specified:
+
+  * Initialize the corresponding repository agent, loading the shared
+    library if necessary. Model loading fails if the shared library is
+    not available or if initialization fails.
+
+  * Invoke the repository agent's *TRITONREPOAGENT_ModelAction*
+    function with action TRITONREPOAGENT_ACTION_LOAD. As input the
+    agent can access the model's repository as either a cloud storage
+    location or a local filesystem location.
+
+  * The repository agent can return *success* to indicate that no
+    changes where made to the repository, can return *failure* to
+    indicate that the model load should fail, or can create a new
+    repository for the model (for example, by decrypting the input
+    repository) and return *success* to indicate that the new
+    repository should be used.
+
+  * If the agent returns *success* Triton continues to the next
+    agent. If the agent returns *failure*, Triton skips invocation of
+    any additional agents.
+
+* If all agents returned *success*, Triton attempts to load the model
+  using the final model repository.
+
+* For each repository agent that was invoked with
+  TRITONREPOAGENT_ACTION_LOAD, in reverse order:
+
+  * Triton invokes the repository agent's
+    *TRITONREPOAGENT_ModelAction* function with action
+    TRITONREPOAGENT_ACTION_LOAD_COMPLETE if the model loaded
+    successfully or TRITONREPOAGENT_ACTION_LOAD_FAIL if the model
+    failed to load.
+
+Triton follows these steps when unloading a model:
+
+* Triton uses the repository agent settings from the initial
+  config.pbtxt file, even if during loading one or more agents
+  modified its contents.
+
+* For each repository agent specified, in reverse order:
```

Comment:
@GuanLuo  I'm going to change this line to "For each repository agent that was invoked with TRITONREPOAGENT_ACTION_LOAD:"
So UNLOAD action will be in forward order, the same as LOAD action